docs/performance/inserts/architecture.txt

```diff
@@ -0,0 +1,263 @@
+
+Single Inserts
+--------------
+
+A single insert looks like this:
+
+.. code-block:: psql
+
+  cr> INSERT INTO my_table (id, first_name)
+  ...      VALUES (1, "Arthur");
+  INSERT OK, 1 row affected  (... sec)
+
+Inserts With Multiple Value Expressions
+---------------------------------------
+
+An insert with :ref:`multiple value expressions <bulk_multiple_values>` looks
+like this:
+
+.. code-block:: psql
+
+   cr> INSERT INTO my_table (id, first_name)
+   ...      VALUES (1, 'Arthur'),
+   ...             (2, 'Trillian'),
+   ...             (2, 'Marvin');
+   INSERT OK, 3 rows affected  (... sec)
+
+Bulk Inserts
+------------
+
+Bulk inserts are typically done via :ref:`the CrateDB HTTP endpoint
+<bulk_switching>`, but they can also be done with :ref:`client batching
+<bulk_client>` or with SQL via :ref:`UNNEST <bulk_unnest>`. Consult the
+respective docs for details of each method.
+
+All three bulk insert methods are handled the same internally by CrateDB.
+
+.. _processing_phases:
+
+Processing Phases
+=================
+
+.. _parsing_analyzing:
+
+Parsing and Analyzing
+---------------------
+
+When a user issues an insert statement to the cluster then this request is sent
+to whichever individual node the client connected to. This node is then
+responsible for performing statement parsing, analyzing, and planning.
+
+The statement is parsed by the *SQL parser*. The *parsed statement* is then
+analyzed by the *statement analyzer*, which produces an *analyzed statement*.
+
+The statement analyzer performs the following checks:
+
+- Does the target table exist?
+
+- Do the specified table columns exist?
+
+   - If not, does the table have a `dynamic column policy`_?
+
+- Do the data types of the insert values match the data types of the existing
+  columns?
+
+- Are all ``PRIMARY KEY`` columns and ``NOT NULL`` column specified?
+
+- Are the values for ``NOT NULL`` columns non null?
+
+- Is the table partitioned?
+
+  - If so, does a `partition`_ exist for all insert values?
+
+    - If not, enrich the statement with a list of the partitions that need to
+      be created
+
+.. _planning:
+
+Planning
+--------
+
+The analyzed statement is then handed off to the *planner*, which plans out the
+execution of the insert. This process continues to take place on the node
+responsible for handling the client request.
+
+How the planner proceeds depends on the type of insert being prepared.
+
+Planning Single Inserts
+.......................
+
+The planner performs the following actions:
+
+1. If the partition doesn't exist, create a new partition
+
+2. Create a single insert request
+
+Planning Inserts With Multiple Value Expressions
+................................................
+
+For each row to be inserted, the planner performs the following actions:
+
+1. If the partition doesn't exist, create a new partition
+
+2. Calculate the shard ID (``ShardId``) using the `routing column`_
+
+3. Create a single insert request for the shard ID
```

Comment:
rows are grouped per shard, so added a shard request (request creation happens lazy underneath if no one exists for that shard yet to be accurate but afaik thats irrelevant here)